### PR TITLE
Issue 414 - 2D Lines support

### PR DESCRIPTION
**Pull request by sebjf**
This addresses #414, to support https://github.com/3drepo/3drepounity/issues/322. It modifies the bouncer to include the primitive type of mesh nodes (falling back to triangles, if none is set). Additionally, the AssetModelExport and MeshMapReorganiser classes have been modified to support face limits and to allow the Unity BouncerWrapper to read the primitive type.

### CODE REVIEW

**Review by carmenfan — CHANGES_REQUESTED** [review]
@sebjf  can we also add some unit tests on the meshnode type? e.g. ensure that repo_bson_factory calculates the correct type given the indices, meshNode will return the right type, etc)

File: bouncer/src/repo/core/model/bson/repo_bson_factory.cpp [hunk 1]
```diff
@@ -393,20 +393,44 @@ MeshNode RepoBSONFactory::makeMeshNode(
...
 			if (!nIndices)
 			{
 				repoWarning << "number of indices in this face is 0!";
 			}
+			if (primitive == MeshNode::Primitive::UNKNOWN)
+			{
+				if (nIndices == 2) {
+					primitive = MeshNode::Primitive::LINES;
+				}
+				if (nIndices == 3) {
```
[the remark this answers is not in the document]
Comment (sebjf) [reply bouncer/src/repo/core/model/bson/repo_bson_factory.cpp]: Corrected in 9b8bc97b59e8b5effd4029ca880fea2883cbf481

File: bouncer/src/repo/core/model/bson/repo_node_mesh.h [hunk 2]
```diff
@@ -112,6 +121,11 @@ namespace repo {
 					return NodeType::MESH;
 				}
 
+				/**
+				* Get the mesh primitive type (points, lines, triangles, quads) as a string (triangles if not set).
```
[the remark this answers is not in the document]
Comment (sebjf) [reply bouncer/src/repo/core/model/bson/repo_node_mesh.h]: Corrected in 9b8bc97b59e8b5effd4029ca880fea2883cbf481

File: bouncer/src/repo/core/model/bson/repo_bson_factory.cpp [hunk 3]
```diff
@@ -393,20 +393,44 @@ MeshNode RepoBSONFactory::makeMeshNode(
...
+					primitive = MeshNode::Primitive::TRIANGLES;
+				}
+				if (primitive == MeshNode::Primitive::UNKNOWN)
```
[the remark this answers is not in the document]
Comment (sebjf) [reply bouncer/src/repo/core/model/bson/repo_bson_factory.cpp]: This end if statement is intentionally separate; this code block attempts to infer the primitive type if it's not already set. The statements above attempt to recover one of two types (that we support, lines or triangles) and the last statement checks if they were successful or not (I've put some comments in a9b6f330aac2e237652aa8e2db8a3225601f7958 for this!)

[the remark this answers is not in the document]
Comment (carmenfan) [reply bouncer/src/repo/core/model/bson/repo_bson_factory.cpp]: the code only enters this part if primitive is unknown, we attempt to set it at L407 and L413. so for the primitive to still be unknown at this point, it means both L407 and L413 were evaluated to false, so this should just be an else?

In other words, there's no situation where L407 or L410 evaluated to true would enter this condition, and there's no way the primitive is set to something other than unknown if they were evaluated to false. 

I don't quite get why it needs to be an if, I feel like I'm missing something 😆

[the remark this answers is not in the document]
Comment (sebjf) [reply bouncer/src/repo/core/model/bson/repo_bson_factory.cpp]: Ah yes you are right! I'll update it...

[the remark this answers is not in the document]
Comment (sebjf) [reply bouncer/src/repo/core/model/bson/repo_bson_factory.cpp]: Took a sec to see past my preconceived design, but updated in 7c13f2e2dc0e840d093d33c9ab2a5cbb74b7c3e1!

**Comment by sebjf** [comment]
@carmenfan I have added some unit tests for the MeshNode with Lines primitives. Let me know how they look/if you want more thorough tests, error cases, etc!

**Review by carmenfan — APPROVED** [review]

**Pull request merged**